### PR TITLE
Allow FileSystem queries to occur from callback threads.

### PR DESCRIPTION
**Pull request by bbockelm**
To query a filesystem property, a free callback thread is needed
with the default synchronous versions of XrdCl::FileSystem methods.

Thus, if the synchronous method is called from a _callback thread_,
then there must be an idle thread to handle the response; if there
are N threads in the pool and N simultaneous queries, then the
filesystem query would cause a deadlock.

Worse yet, the timeout mechanism for XrdCl relies on an idle callback
thread to function.

As making the synchronous query asynchronous in its currrent use case
is quite hard, I only partially solved the problem: wrote a new
response handler that can timeout without an idle thread.

Technically, any version of CMSSW 7_6 or later should be affected.  Pragmatically, do we actually open multiple files simultaneously except in the threaded mixing module?

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

File: Utilities/XrdAdaptor/src/XrdSource.cc [hunk 1]
```diff
@@ -76,6 +76,110 @@ class DelayedClose : boost::noncopyable, public XrdCl::ResponseHandler
...
+        // condvar is notified.
+        std::unique_ptr<char, std::function<void(char*)>> self_ref_guard(nullptr, [&](char *) {m_self.reset();});
+        // On function exit, notify any waiting threads.
+        std::unique_ptr<char, std::function<void(char*)>> notify_guard(nullptr, [&](char *) {m_condvar.notify_all();});
+
+        {
+            // m_mutex protects m_status
+            std::unique_lock<std::mutex> guard(m_mutex);
+            // On exit from the block, make sure m_status is set; it needs to be set before we notify threads.
+            std::unique_ptr<char, std::function<void(char*)>> exit_guard(nullptr, [&](char *) {m_status.reset(new XrdCl::XRootDStatus(XrdCl::stError, XrdCl::errInternal));});
```
Comment (Dr15Jones) [inline Utilities/XrdAdaptor/src/XrdSource.cc]: I didn't think unique_ptr allowed a delete handler.

**Comment by Dr15Jones** [comment]
Why not make ```m_status``` an ```std::atomic<...*>```? If you did that you can use a non ```nullptr``` value to tell you it was sent and ```m_response``` is OK.

**Comment by bbockelm** [comment]
@Dr15Jones - I considered the atomic.  However, we already need to have the mutex there for the condition variable.  Further:
* There's no contention (it's synchronizing between threads).
* There has to be a network round trip anyway, meaning that there will little contention on the mutex.
* Simplifies the implementation of `HandleResponse`.  (Honestly, this was the driving one: with raw pointers, having exception safety got complicated.

File: Utilities/XrdAdaptor/src/XrdSource.cc (hunk 1 above)
Comment (bbockelm) [inline Utilities/XrdAdaptor/src/XrdSource.cc]: Yup, it does!

Since I want to `release()` the data later, I felt it was appropriate here.

**Comment by bbockelm** [comment]
For the static check warnings - I think these are all "correct" and may just need to be annotated.  What are the annotations I should use?

Alternately, there might be some way to tweak the singletons to make the "more correct".  Chris, ideas?

Also, I notice that `catch(...)` is banned by the static analyzer.  I use this to try to make more coherent error messages.  What's the more-correct suggestion?

**Comment by Martin-Grunewald** [comment]
please test
(HIon problem should be solved)

File: Utilities/XrdAdaptor/src/XrdSource.cc [hunk 2]
```diff
@@ -76,6 +76,110 @@ class DelayedClose : boost::noncopyable, public XrdCl::ResponseHandler
...
+class QueryAttrHandler : boost::noncopyable, public XrdCl::ResponseHandler
+{
+public:
+
+    virtual ~QueryAttrHandler() {}
+
+
+    static XrdCl::XRootDStatus query(XrdCl::FileSystem &fs, const std::string &attr, std::chrono::milliseconds timeout, std::string &result)
+    {
+        std::shared_ptr<QueryAttrHandler> handler(new QueryAttrHandler());
```
Comment (Dr15Jones) [inline Utilities/XrdAdaptor/src/XrdSource.cc]: Why isn't this just on the stack?

Comment (Dr15Jones) [inline Utilities/XrdAdaptor/src/XrdSource.cc]: I see now, you made it own itself

File: Utilities/XrdAdaptor/src/XrdSource.cc [hunk 3]
```diff
@@ -76,6 +76,110 @@ class DelayedClose : boost::noncopyable, public XrdCl::ResponseHandler
...
+        XrdCl::Buffer arg(attr.size());
+        arg.FromString(attr);
+
+        // On error or exception thrown, drop the reference to ourself
+        std::unique_ptr<QueryAttrHandler, std::function<void(QueryAttrHandler*)>> self_ref_guard(nullptr, [&](QueryAttrHandler*) {handler->m_self.reset();});
```
Comment (Dr15Jones) [inline Utilities/XrdAdaptor/src/XrdSource.cc]: Why not pull lines 99 to 110 within a block. That way the self_ref_guard usage is clearly deliniated.

File: Utilities/XrdAdaptor/src/XrdSource.cc [hunk 4]
```diff
@@ -76,6 +76,110 @@ class DelayedClose : boost::noncopyable, public XrdCl::ResponseHandler
...
+        XrdCl::XRootDStatus st = fs.Query(XrdCl::QueryCode::Config, arg, handler.get());
+        if (!st.IsOK())
+        {
+            return st;
+        }
+
+        // Successfully registered the callback; keep the self-reference until the
+        // callback release it.
+        self_ref_guard.release();
+
```
Comment (Dr15Jones) [inline Utilities/XrdAdaptor/src/XrdSource.cc]: OK, at this point you will have the handler maintain a circular dependency on itself. How does it get deleted?

**Comment by Dr15Jones** [comment]
I think the memory management can be achieved more simply and avoid the QueryAttrHandler holding a std::shared_ptr to itself.

The ```QueryAttrHandler``` is really only used for setting the values of m_status and m_response which is all the ```query``` method wants. If the timeout happens we do not care about those values.

So instead, create ```std::shared_ptr``` on the stack for ```status``` and ```response``` and then pass ```std::weak_ptr``` from those smart pointers to the constructor of ```QueryAttrHandler```. When the handler is calledback from xrootd if there was no timeout the conversion of ```std::weak_ptr``` to ```std::shared_ptr``` will succeed and you can do your work normally. If the timeout happened one or both conversions will fail and you will know not to proceed with the work. We then just have xrootd delete the handler.

**Comment by bbockelm** [comment]
Hi Chris,

I went ahead with your suggestion but found the resulting code was much harder to follow (to me).  I also had to wrap m_mutex to keep it alive when the callback occurred before timeout.

So, a single self-reference was replaced by three `std::weak_ptr`; it didn't seem to buy us much!

Brian

**Comment by davidlange6** [comment]
please test

**Comment by bbockelm** [comment]
The update fixes a the obvious build issue (sorry!  refactored a line after testing... guess which one) and attacks the refactor suggested by Chris.

**Comment by Dr15Jones** [comment]
please test

File: Utilities/XrdAdaptor/src/XrdSource.cc [hunk 5]
```diff
@@ -246,9 +355,7 @@ Source::getXrootdSiteFromURL(std::string url, std::string &site)
         delete response;
         return false;
     }
-    std::string rsite = response->ToString();
-    delete response;
```
Comment (Dr15Jones) [inline Utilities/XrdAdaptor/src/XrdSource.cc]: It looks like ```response``` is an unused variable?

**Comment by Dr15Jones** [comment]
+1
Any further changes can be done on a different pull request since this will fix a problem which is presently 'deadlocking' MC jobs in the IB RelVals for CMSSW_8_0_THREADED_X.

**Comment by Dr15Jones** [comment]
Upon further reflection, I don't think this will solve the underlying problem and will only push the problem to a slightly later stage.

What is happening is xrootd only has 3 threads it uses to answer client 'queries'. In our case each of those threads gets stuck with the following stacktrace

```
#0  0x00000037912e5199 in syscall () from /lib64/libc.so.6
#1  0x00007fab05781500 in XrdSys::LinuxSemaphore::Wait() () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/external/slc6_amd64_gcc493/lib/libXrdCl.so.2
#2  0x00007fab0577609a in XrdCl::FileSystem::Query(XrdCl::QueryCode::Code, XrdCl::Buffer const&, XrdCl::Buffer*&, unsigned short) () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/external/slc6_amd64_gcc493/lib/libXrdCl.so.2
#3  0x00007fab05702d70 in XrdAdaptor::Source::getXrootdSiteFromURL(std::string, std::string&) () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/lib/slc6_amd64_gcc493/libUtilitiesXrdAdaptor.so
#4  0x00007fab0570448e in XrdAdaptor::Source::getXrootdSite(XrdCl::File&, std::string&) () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/lib/slc6_amd64_gcc493/libUtilitiesXrdAdaptor.so
#5  0x00007fab05704a40 in XrdAdaptor::Source::setXrootdSite() () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/lib/slc6_amd64_gcc493/libUtilitiesXrdAdaptor.so
#6  0x00007fab05705da5 in XrdAdaptor::Source::Source(timespec, std::unique_ptr<XrdCl::File, std::default_delete<XrdCl::File> >, std::string const&) () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/lib/slc6_amd64_gcc493/libUtilitiesXrdAdaptor.so
#7  0x00007fab056f3822 in XrdAdaptor::RequestManager::OpenHandler::HandleResponseWithHosts(XrdCl::XRootDStatus*, XrdCl::AnyObject*, std::vector<XrdCl::HostInfo, std::allocator<XrdCl::HostInfo> >*) () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/lib/slc6_amd64_gcc493/libUtilitiesXrdAdaptor.so
#8  0x00007fab057a2dfc in ?? () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/external/slc6_amd64_gcc493/lib/libXrdCl.so.2
#9  0x00007fab057860cf in XrdCl::XRootDMsgHandler::HandleResponse() () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/external/slc6_amd64_gcc493/lib/libXrdCl.so.2
#10 0x00007fab05787ffe in XrdCl::XRootDMsgHandler::Process(XrdCl::Message*) () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/external/slc6_amd64_gcc493/lib/libXrdCl.so.2
#11 0x00007fab0576b20a in XrdCl::Stream::HandleIncMsgJob::Run(void*) () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/external/slc6_amd64_gcc493/lib/libXrdCl.so.2
#12 0x00007fab057cbc16 in XrdCl::JobManager::RunJobs() () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/external/slc6_amd64_gcc493/lib/libXrdCl.so.2
#13 0x00007fab057cbdf9 in ?? () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/external/slc6_amd64_gcc493/lib/libXrdCl.so.2
#14 0x0000003791607a51 in start_thread () from /lib64/libpthread.so.0
#15 0x00000037912e893d in clone () from /lib64/libc.so.6
```

The reason they are stuck is ```XrdCl::FileSystem::Query``` puts another client ```query``` on xrootd's list of queries to run but that new query can't run because all the threads for queries are stuck waiting for a query that is on the list but can't run because the threads are all full.

This pull request allows the threads mangaged by cmsRun (i.e. the TBB threads) to eventually timeout waiting for xrootd to respond and then proceed to run. However, all subsequent calls will also time out since from this point on in the program xrootd will never be able to give a client response! Eventually we'll try to open a new file, it will time out which will lead to an exception which will abort the job.

The real fix we need is to not get xrootd's query threads jammed up in the first place.  I'll start looking into how to accomplish that task.

**Comment by bbockelm** [comment]
> This pull request allows the threads mangaged by cmsRun (i.e. the TBB threads) to eventually timeout waiting for xrootd to respond and then proceed to run

Nope.  This fix affects this frame:
```
#3  0x00007fab05702d70 in XrdAdaptor::Source::getXrootdSiteFromURL(std::string, std::string&) () from /afs/cern.ch/cms/sw/ReleaseCandidates/vol1/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_THREADED_X_2016-01-28-1100/lib/slc6_amd64_gcc493/libUtilitiesXrdAdaptor.so
```

I.e., the change is that `getXrootdSiteFromURL` will timeout without needing an XRootD callback.  Hence, the `OpenHandler` callback will not block, finish, and free up the XRootD callback.

This is all for the XRootD callback threads; there shouldn't be changes for the TBB-managed threads.

**Comment by Dr15Jones** [comment]
Ah, it is because you now call the 'asynchronous' version of ```XrdCl::FileSystem::Query``` rather than the synchronous version with the same name. It was the same name that lead to my confusion when I looked at the stacktrace again. Man, I wish they would have called it  ```XrdCl::FileSystem::AsyncQuery```.

Sorry for the false alarm.

**Pull request merged**